components/raftstore/src/store/peer.rs

```diff
@@ -3556,19 +3583,92 @@ where
                 cmd_type
             ));
         }
-        if entry_size as f64 > ctx.cfg.raft_entry_max_size.0 as f64 * 0.9 {
+        let entry_size_limit = ctx.cfg.raft_entry_max_size.0 as usize * 9 / 10;
+        if entry_size > entry_size_limit {
             return Err(box_err!(
                 "log gap size exceed entry size limit, skip merging."
             ));
+        };
+
+        // Record current proposed index. If there are some in-memory pessimistic locks, we should
+        // wait until applying to the proposed index before proposing pessimistic locks and
+        // PrepareMerge. Otherwise, if an already proposed command will remove a pessimistic lock,
+        // we will make some deleted locks appear again.
+        if !passed_merge_fence {
+            let pessimistic_locks = self.txn_ext.pessimistic_locks.read();
+            if !pessimistic_locks.is_empty() {
+                if !pessimistic_locks.is_valid {
```

Comment:
Could we assert it should always be true because all cases have already been checked?

Reply:
I think it's possible that it's going to have a transfer leader, so `is_valid` becomes false. And related commands also have been committed, so the checks are pased.

Reply:
Can you add some comments for clarifying this?